src/main/java/fi/okm/jod/yksilo/config/elasticache/IamAuthTokenRequest.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2024 The Finnish Ministry of Education and Culture, The Finnish
+ * The Ministry of Economic Affairs and Employment, The Finnish National Agency of
+ * Education (Opetushallitus) and The Finnish Development and Administration centre
+ * for ELY Centres and TE Offices (KEHA).
+ *
+ * Licensed under the EUPL-1.2-or-later.
+ */
+
+package fi.okm.jod.yksilo.config.elasticache;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Collections;
+import software.amazon.awssdk.auth.credentials.AwsCredentials;
+import software.amazon.awssdk.auth.signer.Aws4Signer;
+import software.amazon.awssdk.auth.signer.params.Aws4PresignerParams;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.http.SdkHttpMethod;
+import software.amazon.awssdk.regions.Region;
+
+public class IamAuthTokenRequest {
+  private static final String REQUEST_PROTOCOL = "http";
+  private static final String REQUEST_PROTOCOL_WITH_COLON_AND_SLASHES = REQUEST_PROTOCOL + "://";
+  private static final String REQUEST_PATH = "/";
+  private static final String PARAM_ACTION = "Action";
+  private static final String PARAM_USER = "User";
+  private static final String PARAM_RESOURCE_TYPE = "ResourceType";
+  private static final String RESOURCE_TYPE_SERVERLESS_CACHE = "ServerlessCache";
+  private static final String ACTION_NAME = "connect";
+  private static final String SERVICE_NAME = "elasticache";
+  private static final Duration TOKEN_EXPIRY_DURATION_SECONDS = Duration.ofSeconds(900);
+
+  private final String userId;
+  private final String cacheName;
+  private final Region region;
+
+  public IamAuthTokenRequest(String userId, String cacheName, Region region) {
+    this.userId = userId;
+    this.cacheName = cacheName;
+    this.region = region;
+  }
+
+  public String toSignedRequestUri(AwsCredentials credentials) {
+    SdkHttpFullRequest request = getSignableRequest();
+    request = sign(request, credentials);
+    return request.getUri().toString().replace(REQUEST_PROTOCOL_WITH_COLON_AND_SLASHES, "");
+  }
+
+  private SdkHttpFullRequest getSignableRequest() {
+    return SdkHttpFullRequest.builder()
+        .method(SdkHttpMethod.GET)
+        .protocol(REQUEST_PROTOCOL)
+        .host(cacheName)
+        .encodedPath(REQUEST_PATH)
+        .appendRawQueryParameter(PARAM_ACTION, ACTION_NAME)
+        .appendRawQueryParameter(PARAM_USER, userId)
+        .putRawQueryParameter(
+            PARAM_RESOURCE_TYPE, Collections.singletonList(RESOURCE_TYPE_SERVERLESS_CACHE))
+        .build();
+  }
+
+  private SdkHttpFullRequest sign(SdkHttpFullRequest request, AwsCredentials credentials) {
+    Instant expiryInstant = Instant.now().plus(TOKEN_EXPIRY_DURATION_SECONDS);
+    Aws4Signer signer = Aws4Signer.create();
```

Comment:
Signer could be also created just once in the constructor (for the same reason as eliminating the SignableRequest)
(see e.g. https://github.com/aws/aws-sdk-java-v2/blob/master/services/rds/src/main/java/software/amazon/awssdk/services/rds/DefaultRdsUtilities.java)